bokeh/models/renderers.py

```diff
@@ -189,42 +192,66 @@ def __init__(self, **kw):
 
     level = Override(default="glyph")
 
-_DEFAULT_NODE = lambda: Circle()
+_DEFAULT_NODE_RENDERER = lambda: GlyphRenderer(
+    glyph=Circle(), data_source=ColumnDataSource(data=dict(index=[]))
+)
 
-_DEFAULT_EDGE = lambda: MultiLine()
+_DEFAULT_EDGE_RENDERER = lambda: GlyphRenderer(
+    glyph=MultiLine(), data_source=ColumnDataSource(data=dict(start=[], end=[]))
+)
 
 class GraphRenderer(DataRenderer):
     '''
 
     '''
 
+    @error(MALFORMED_GRAPH_SOURCE)
+    def _check_malformed_graph_source(self):
+        missing = []
+        if "index" not in self.node_renderer.data_source.column_names:
+            missing.append("Column 'index' is missing in GraphSource.node_renderer.data_source")
+        if "start" not in self.edge_renderer.data_source.column_names:
+            missing.append("Column 'start' is missing in GraphSource.edge_renderer.data_source")
+        if "end" not in self.edge_renderer.data_source.column_names:
+            missing.append("Column 'end' is missing in GraphSource.edge_renderer.data_source")
+
+        indexes = self.node_renderer.data_source.data.get('index', [])
+        for i in self.edge_renderer.data_source.data.get('start', []):
+            if i not in indexes:
+                missing.append("GraphSource.edge_renderer.data_source 'start' value '%s' missing from GraphSource.node_renderer.data_source 'index' column" % i)
+
+        for i in self.edge_renderer.data_source.data.get('end', []):
+            if i not in indexes:
+                missing.append("GraphSource.edge_renderer.data_source 'end' value '%s' missing from GraphSource.node_renderer.data_source 'index' column" % i)
```

Comment:
These checks seems potentially very expensive? I guess we more usually we validate the overall format and arrangements of data without delving into the values (categorical factors with ":" being one exception). I might suggest removing these for now, and making an issue to add them separately (but also to be able to configure the validation system to turn various checks on or off which is currently not possible)